### PR TITLE
Add a meritocracy

### PR DESCRIPTION
**Pull request by PlasmaPower**
I decided to not go with a clickbait title this time :smile:

Look at #361 for more details, but **I tried to give the meritocracy minimal power. Democracy still decides the outcome >90% of the time.** This is just to prevent against sneaky take over attempts.

Basis of how it works:

> At least one positive non-author meritocracy review on the most recent commit of a PR is required for that PR to be accepted.

Changes from #361:

- Autogenerate meritocracy list (currently top 10 contributors + top 10 voters)
- Don't count reviews as votes (so members of the meritocracy can allow a PR without voting for it)

### CODE REVIEW

**Comment by mark-i-m** [comment]
For anyone else looking through this PR, it's a lot easier to read if you append `?w=1` to the URL, which gets rid of whitespace changes (e.g. indentation changes).

**Comment by PlasmaPower** [comment]
@mark-i-m yes definitely do that

*hopes nobody notices my trojan horse coded in Whitespace*

File: github_api/voting.py [hunk 1]
```diff
@@ -26,15 +27,16 @@ def get_votes(api, urn, pr):
         votes[voter] = vote
 
     # get all the pr-review-based votes
-    for vote_owner, vote in get_pr_review_votes(api, urn, pr_num):
-        if vote and vote_owner != pr_owner:
-            votes[vote_owner] = vote
+    for vote_owner, is_current, vote in get_pr_review_votes(api, urn, pr):
+        if (vote > 0 and is_current and vote_owner != pr_owner
```
Comment (mark-i-m) [inline github_api/voting.py:31]: Should this be `vote_owner.lower() != pr_owner.lower()`? i.e. are we sure the cases match?

**Comment by phil-r** [comment]
👍 What can possibly go wrong?

File: github_api/voting.py (hunk 1 above)
Comment (PlasmaPower) [reply github_api/voting.py:31]: That's how it was, and it works since they both come from the GitHub API.

Comment (mark-i-m) [reply github_api/voting.py:31]: Ok, that's what I thought

**Comment by PlasmaPower** [comment]
@phil-r your comment keeps duplicating client-side:

screenshot

No idea why. It seems to be a GitHub bug. The bug survives a refresh, although the duplicated comment versions don't.

Edit: posting another comment seems to have fixed it.

**Comment by phil-r** [comment]
@PlasmaPower I've also noticed, refresh helped, maybe I've triggered some easter egg? 🐰 🥚

**Comment by PlasmaPower** [comment]
@phil-r I think it's actually a bug with reviews. It only seems to happen when the comment is next to a review. Overall, reviews are poorly implemented IMO.

**Comment by PlasmaPower** [comment]
When you fix a minor mistake but it resets the voting window :cry:

I guess #378 wins the merge conflict race.

**Comment by mark-i-m** [comment]
Hey! :tada:

**Comment by mark-i-m** [comment]
What was the fix?

**Comment by PlasmaPower** [comment]
@mark-i-m In `patch.py` I cached `gh_api.prs.get_contributors` not `gh_api.repos.get_contributors`.

**Comment by mark-i-m** [comment]
Btw @PlasmaPower you down voted your PR by crying in the post above...

**Comment by PlasmaPower** [comment]
@mark-i-m lol oops. :+1: I think that fixes it?

Edit: yep

**Comment by PlasmaPower** [comment]
Reminds me of when reactions to comments counted for the PR as a whole.

**Comment by PlasmaPower** [comment]
Had to do another slight fix: `.iteritems()` was removed from Python 3. Changed to just `.items()`

**Comment by mark-i-m** [comment]
@PlasmaPower Are you squashing the commits in your own fork?

**Comment by PlasmaPower** [comment]
@mark-i-m Oh yeah, I probably shouldn't now that chaosbot squashes them for us. Doesn't make much of a difference though, and there's no added security - I could always amend a malicious change to a previous commit.

**Comment by mark-i-m** [comment]
I have my :eye: on you

**Comment by PlasmaPower** [comment]
*waits until @mark-i-m goes to sleep* /s

**Comment by PlasmaPower** [comment]
Or do you sleep with that eye open?

**Comment by mark-i-m** [comment]
You don't know if I sleep (at the same time as you)

**Comment by PlasmaPower** [comment]
I just realized I forgot to pass `reverse=True` to sort, so it was an opposite meritocracy... oops! Will be fixed soon, just double checking a few things.

**Comment by davidak** [comment]
@PlasmaPower this solution looks very good!

So at least one member of this group has to do a review seperately from voting. So people needs to know if they are in this group and that it's helpful if they review. Also that they have responsibility for the project doing it.

It would be good if you add that to the readme and we need some kind of news section to inform the community of such important changes.

**Comment by davidak** [comment]
@PlasmaPower If this get's merged you can mention everyone currently in the list so they know that they now have this opportunity.

**Comment by PlasmaPower** [comment]
Current meritocracy: @chaosbot, @muchzer, @pchauncey, @amoffat, @hongaar, @rhengles, @rudehn, @xyproto, @eukaryote31, @ad-m, @mark-i-m, @smittyvb, @reddraggone9, @plasmapower, @phil-r

I've updated the README, pushing the changes in a second. The meritocracy will be logged every time chaosbot checks for new PRs.

(yes I know chaosbot is in the list, but hardcoding chaosbot is both unnecessary and probably a bad idea. chaosbot will also fade from the list as merge commits give way to squashes)

**Comment by PlasmaPower** [comment]
Okay, both of those changes have been pushed. Ready to merge! :tada:

Other slight updates can wait for another PR, I'll only add on important fixes.

(now to wait 9 hours because the extended voting window is broken... :sleeping:)

**Comment by davidak** [comment]
>Current meritocracy

@PlasmaPower shoudn't that be 20 people?

**Comment by andrewda** [comment]
@davidak some people are duplicates – they're in both the top 10 contributors and the top 10 voters

**Comment by andrewda** [comment]
We should eventually have a webpage up similar to http://chaosthebot.com/voters.html to show the current meritocracy. I can work on that tomorrow.

**Comment by PlasmaPower** [comment]
@andrewda I agree, the reason I didn't implement it now is because it'd be the first dynamic part of the server.

**Comment by PlasmaPower** [comment]
Rebased - fixed `settings.py` conflicts.

**Comment by PlasmaPower** [comment]
The Travis failure appears to be from master.

Edit: yep, fixed in #390

**Comment by andrewda** [comment]
@PlasmaPower 
> @andrewda I agree, the reason I didn't implement it now is because it'd be the first dynamic part of the server.

It doesn't have to be dynamic. How I currently have it written is to update a `meritocracy.json` every time it's generated, just like how the `voters.json` works. Both approaches are a little bit gross, but we can work on refactoring that and/or using Django in another PR.

**Comment by tarunbatra** [comment]
This is something nice, @PlasmaPower. Though it chooses the same set of ppl, more or less, but look at the response. 😄

**Comment by PlasmaPower** [comment]
@tarunbatra It's also nice that it automatically updates itself. I might adjust the config later though.

**Comment by PlasmaPower** [comment]
> How I currently have it written is to update a meritocracy.json every time it's generated, just like how the voters.json works.

The difference there is that `voters.json` needs to survive restarts, but `meritocracy.json` doesn't (it'd be essentially write-only). I think a dynamic server is the better solution.

**Comment by andrewda** [comment]
I completely agree

**Comment by hongaar** [comment]
well well well... another regime change by @PlasmaPower!

**Comment by andrewda** [comment]
For some reason, this duplicated the object in [`voters.json`](http://chaosthebot.com/voters.json) and now it's invalid. This breaks the [voters page](http://chaosthebot.com/voters.html).

```json
{ "amoffat": 12468, "g19fanatic": 1442, ... }{ "amoffat": 12468, "g19fanatic": 1442, ... }
```

**Comment by PlasmaPower** [comment]
That might also remove the top 10 voters part of the meritocracy. Anybody know what's causing it?

**Comment by andrewda** [comment]
It's also crashing repeatedly http://chaosthebot.com:8081/chaos-stderr.log

**Comment by PlasmaPower** [comment]
Oh I think that's caused by two PRs being merged at the same time. I'll get a fix ready.

**Comment by rudehn** [comment]
So what happens when somebody who is in the meritocracy reviews the PR, but falls out of the meritocracy before the PR is merged? Does the review no longer count?

**Comment by andrewda** [comment]
I believe that the next time the PR gets polled, the review will no longer count.

**Comment by rudehn** [comment]
Is that working as intended, or a bug?

**Comment by PlasmaPower** [comment]
That's how it works, and there's really no way to fix it unless we get the database up and running (even then it'd be a bit of a pain).

**Comment by mdcfe** [comment]
@PlasmaPower This *is* supposed to be chaos - who knows?

**Pull request merged**